Issue#35: Users CRUD

mykytam opened the pull request:
dev
## ZenHub

* [Main ZenHub ticket](https://app.zenhub.com/workspaces/home-project-5f7b77ff8db73a001cb17009/issues/ita-social-projects/home/35)


## Code reviewers

- [x] @alexorlenko
- [x] @Arthurpanda  
- [x] @mykytam 
- [x] @omelchenkotimur 
- [ ] @Prosperro-de 
- [x] @reeztl 

### Second Level Review

- [ ] @DzigMS 
- [ ] @kh0ma 

## Summary of issue

Add User's CRUD operations

## Summary of change

As anonymous user I want to create new account
As user I want to update/delete my account
As admin I want to create/read/update/delete some user

## Testing approach

ToDo

## CHECK LIST
- [ ]  СI passed
- [ ]  Сode coverage >=95%
- [ ]  PR is reviewed manually again (to make sure you have 100% ready code)
- [ ]  All reviewers agreed to merge the PR
- [ ]  I've checked new feature as logged in and logged out user if needed
- [ ]  PR meets all conventions

mykytam committed:
Merge remote-tracking branch 'origin/Issue#34-Create-User-entity-and-configure-Auth-problem' into Issue#35-Users-CRUD

mykytam committed:
Issue#35: Users CRUD
Initial commit
Add User services

mykytam committed:
Merge remote-tracking branch 'origin/Issue#59-Refactor-OpenApi-spec' into Issue#35-Users-CRUD

mykytam committed:
Merge remote-tracking branch 'origin/Issue#59-Refactor-OpenApi-spec' into Issue#35-Users-CRUD

mykytam committed:
Issue#35: Users CRUD
Delete interface for services
Update services

mykytam committed:
Merge remote-tracking branch 'origin/Issue#34-Create-User-entity-and-configure-Auth-problem' into Issue#35-Users-CRUD

# Conflicts:
#	home-data/src/main/java/com/softserveinc/ita/homeproject/homedata/entity/Role.java
#	home-data/src/main/java/com/softserveinc/ita/homeproject/homedata/entity/User.java

mykytam committed:
Issue#35: Users CRUD
Change return type of delete operation

mykytam committed:
Merge remote-tracking branch 'origin/dev' into Issue#35-Users-CRUD

# Conflicts:
#	home-data/src/main/resources/application.properties

mykytam committed:
Merge remote-tracking branch 'origin/dev' into Issue#35-Users-CRUD

kh0ma commented on home-application/src/main/java/com/softserveinc/ita/homeproject/application/apiservice/UserApiServiceImpl.java:
```diff
@@ -0,0 +1,53 @@
+@RequiredArgsConstructor
+public class UserApiServiceImpl implements UsersApiService {
+
+    private final UserServiceImpl userServiceImpl;
```
Please use the interface instead of implementation.

kh0ma replied:
...

mykytam replied:
Sure, thanks

kh0ma commented on home-application/src/main/java/com/softserveinc/ita/homeproject/application/apiservice/UserApiServiceImpl.java:
```diff
@@ -0,0 +1,53 @@
+
+    @PreAuthorize("hasAuthority('CREATE_USER_PERMISSION')")
+    @Override
+    public Response createUser(CreateUser createUser, SecurityContext securityContext) throws NotFoundException {
```
How `throws NotFoundException` can appear here?
Can we get rid of `SecurityContext securityContext`?

mykytam replied:
Ok, noted, I will try

kh0ma commented on home-application/src/main/java/com/softserveinc/ita/homeproject/application/apiservice/UserApiServiceImpl.java:
```diff
@@ -0,0 +1,53 @@
+
+    private final UserServiceImpl userServiceImpl;
+
+    @PreAuthorize("hasAuthority('CREATE_USER_PERMISSION')")
```
Is it possible to move `CREATE_USER_PERMISSION` to constants?

mykytam replied:
Yes, sure. Noted

kh0ma commented on home-application/src/main/java/com/softserveinc/ita/homeproject/application/apiservice/UserApiServiceImpl.java:
```diff
@@ -0,0 +1,53 @@
+
+    @PreAuthorize("hasAuthority('UPDATE_USER_PERMISSION')")
+    @Override
+    public Response updateUser(Integer id, UpdateUser updateUser, SecurityContext securityContext) throws NotFoundException {
```
Do we really use `Integer` for id? I think it should be `Long`.

mykytam replied:
Generator generates Integer value, I will look into it and see if it's possible to change it

mykytam replied:
I will fix it

kh0ma commented on home-application/src/main/java/com/softserveinc/ita/homeproject/application/apiservice/UserApiServiceImpl.java:
```diff
@@ -0,0 +1,53 @@
+
+    @PreAuthorize("hasAuthority('GET_USERS_PERMISSION')")
+    @Override
+    public Response usersGet(@Min(0) Integer offset, @Min(1) @Max(100) Integer limit, SecurityContext securityContext) throws NotFoundException {
```
Please properly name the method. What about `getUsers` or `queryUsers`?

kh0ma replied:
Add sorting.

kh0ma replied:
Pagination has not really good developer experience. Please consider using page size and page number as query parameters and total elements, page size, current page as headers in a response.

mykytam replied:
About namig: plugin generates interface with this naming, I will try to find out how we could name it ourselves
Soring&pagination: noted

kh0ma commented on home-application/src/main/java/com/softserveinc/ita/homeproject/application/apiservice/UserApiServiceImpl.java:
```diff
@@ -0,0 +1,53 @@
+    @PreAuthorize("hasAuthority('GET_USERS_PERMISSION')")
+    @Override
+    public Response usersGet(@Min(0) Integer offset, @Min(1) @Max(100) Integer limit, SecurityContext securityContext) throws NotFoundException {
+        return Response.ok().entity(userServiceImpl.getAllUsers()).build();
```
Pagination isn't applied here.

mykytam replied:
Noted

kh0ma commented on home-application/src/main/java/com/softserveinc/ita/homeproject/application/apiservice/UserApiServiceImpl.java:
```diff
@@ -0,0 +1,53 @@
+    @Override
+    public Response usersIdDelete(Integer id, SecurityContext securityContext) throws NotFoundException {
+        userServiceImpl.deactivateUser(id.longValue());
+        return Response.ok().entity(Response.status(Response.Status.OK)).build();
```
Why do we need to pass entity here?

kh0ma replied:
Does response status match the contract?

mykytam replied:
No, we don't need entity in this response, noted, thanks

kh0ma commented on home-application/src/main/java/com/softserveinc/ita/homeproject/application/apiservice/UserApiServiceImpl.java:
```diff
@@ -0,0 +1,53 @@
+
+    @PreAuthorize("hasAuthority('DELETE_USER_PERMISSION')")
+    @Override
+    public Response usersIdDelete(Integer id, SecurityContext securityContext) throws NotFoundException {
```
Method naming. Consider `removeUser`.

mykytam replied:
Noted, naming will be fixed

kh0ma commented on home-application/src/main/java/com/softserveinc/ita/homeproject/application/apiservice/UserApiServiceImpl.java:
```diff
@@ -0,0 +1,53 @@
+    @PreAuthorize("hasAuthority('GET_USER_PERMISSION')")
+    @Override
+    public Response usersIdGet(Integer id, SecurityContext securityContext) throws NotFoundException {
+        return Response.ok().entity(userServiceImpl.getUserByParameter(id.longValue())).build();
```
How does it work?
```
userServiceImpl.getUserByParameter(id.longValue())
```

mykytam replied:
Method returns the value of this Integer as a long. It's needed because of generated id type

mykytam replied:
I will try to change type to Long

kh0ma commented on home-application/src/main/java/com/softserveinc/ita/homeproject/application/apiservice/UserApiServiceImpl.java:
```diff
@@ -0,0 +1,53 @@
+    @PreAuthorize("hasAuthority('CREATE_USER_PERMISSION')")
+    @Override
+    public Response createUser(CreateUser createUser, SecurityContext securityContext) throws NotFoundException {
+        return Response.ok().entity(userServiceImpl.createUser(createUser)).build();
```
Does response status match the contract?

mykytam replied:
You mean we should use here one more dto layer?

kh0ma commented on home-application/src/main/java/com/softserveinc/ita/homeproject/application/apiservice/UserApiServiceImpl.java:
```diff
@@ -0,0 +1,53 @@
+
+    @PreAuthorize("hasAuthority('GET_USER_PERMISSION')")
+    @Override
+    public Response usersIdGet(Integer id, SecurityContext securityContext) throws NotFoundException {
```
Method naming. `getUser`

mykytam replied:
Noted, naming will be fixed.

kh0ma commented on home-data/src/main/java/com/softserveinc/ita/homeproject/homedata/entity/Permission.java:
```diff
@@ -26,7 +27,7 @@
     private Long id;
 
-    @ManyToMany(mappedBy = "permissions")
+    @ManyToMany(mappedBy = "permissions", fetch = FetchType.LAZY)
```
Isn't lazy the default one?

mykytam replied:
Yes, it is. I can remove it, if it's critical

kh0ma commented on home-service/src/main/java/com/softserveinc/ita/homeproject/service/UserService.java:
```diff
@@ -0,0 +1,20 @@
+package com.softserveinc.ita.homeproject.service;
+
+import com.softserveinc.ita.homeproject.model.CreateUser;
```
The package isn't clear:
```
com.softserveinc.ita.homeproject.model.CreateUser
```

I think it should be `com.softserveinc.ita.homeproject.api.model.CreateUser`

kh0ma replied:
Why do we depend on api from service?

mykytam replied:
It's the package, where we have all our DTOs etc.

kh0ma commented on home-service/src/main/java/com/softserveinc/ita/homeproject/service/UserService.java:
```diff
@@ -0,0 +1,20 @@
+
+public interface UserService {
+
+    ReadUser createUser(CreateUser payload);
```
Use DTOs.

mykytam replied:
Are `ReadUser` and `CreateUser` not generated DTOs?

kh0ma commented on home-service/src/main/java/com/softserveinc/ita/homeproject/service/UserService.java:
```diff
@@ -0,0 +1,20 @@
+
+    ReadUser createUser(CreateUser payload);
+
+    ReadUser updateUser(Long id, UpdateUser payload);
```
avoid such naming - `payload`. Would be better `UpdateUserDto updateUserDto`

mykytam replied:
Noted, will be fixed

kh0ma commented on home-service/src/main/java/com/softserveinc/ita/homeproject/service/UserService.java:
```diff
@@ -0,0 +1,20 @@
+
+    Collection<ReadUser> getAllUsers();
+
+    ReadUser getUserByParameter(Long id);
```
By which parameter?

mykytam replied:
On one of the meeting you said that it's better to use not byId, but more general concept

kh0ma commented on home-service/src/main/java/com/softserveinc/ita/homeproject/service/impl/UserServiceImpl.java:
```diff
@@ -0,0 +1,75 @@
+    @Override
+    public ReadUser createUser(CreateUser payload) {
+        if (userRepository.findByEmail(payload.getEmail()).isPresent()) {
+            throw new RuntimeException();  //TODO: here throw already exist exception
```
Resolve TODO

mykytam replied:
Should I make "exceptions" task in this branch?

kh0ma commented on home-service/src/main/java/com/softserveinc/ita/homeproject/service/impl/UserServiceImpl.java:
```diff
@@ -0,0 +1,75 @@
+        if (userRepository.findByEmail(payload.getEmail()).isPresent()) {
+            throw new RuntimeException();  //TODO: here throw already exist exception
+        } else {
+            User toCreate = conversionService.convert(payload, User.class);
```
avoid such naming

mykytam replied:
Ok, noted

kh0ma commented on home-service/src/main/java/com/softserveinc/ita/homeproject/service/impl/UserServiceImpl.java:
```diff
@@ -0,0 +1,75 @@
+            User toCreate = conversionService.convert(payload, User.class);
+            toCreate.setEnabled(true);
+            toCreate.setExpired(false);
+            toCreate.setRoles(Set.of(roleRepository.findByName("USER")));
```
Is `USER` the default role for all newly created users? Please move it to constant.

mykytam replied:
Ok, noted, thanks

kh0ma commented on home-service/src/main/java/com/softserveinc/ita/homeproject/service/impl/UserServiceImpl.java:
```diff
@@ -0,0 +1,75 @@
+            userRepository.save(toUpdate);
+            return conversionService.convert(payload, ReadUser.class);
+        } else {
+            throw new RuntimeException();  //TODO: here throw not found exception
```
Resolve it.

mykytam replied:
Same as previous, isn't it the point of other task?

mykytam committed:
Issue#35: Users CRUD
Delete redundant fetchType from models
Use the interface instead of implementation in api service
Add constants for Roles and Permissions
Change payload's name

mykytam committed:
Merge remote-tracking branch 'origin/dev' into Issue#35-Users-CRUD

mykytam committed:
Issue#35: Users CRUD
Relocate constants
Add DTO layer for api services
Add exceptions
Change models in services

mykytam committed:
Issue#35: Users CRUD
Change returning variable

mykytam committed:
Issue#35: Users CRUD
Change api base error class

kh0ma commented on home-application/src/main/java/com/softserveinc/ita/homeproject/application/config/JerseyConfig.java:
```diff
@@ -1,5 +1,6 @@
 package com.softserveinc.ita.homeproject.application.config;
 
+
```
extra line

mykytam replied:
I will delete it, thanks

kh0ma commented on home-data/src/main/java/com/softserveinc/ita/homeproject/homedata/entity/Permission.java:
```diff
@@ -12,6 +12,7 @@
 import javax.persistence.Id;
 import javax.persistence.Column;
 import javax.persistence.ManyToMany;
+import javax.persistence.FetchType;
```
Isn't this one redundant?

mykytam replied:
Yes, it is, thanks

kh0ma commented on home-data/src/main/java/com/softserveinc/ita/homeproject/homedata/entity/Role.java:
```diff
@@ -14,6 +14,7 @@
 import javax.persistence.ManyToMany;
 import javax.persistence.JoinTable;
 import javax.persistence.JoinColumn;
+import javax.persistence.FetchType;
```
Isn't this one redundant?

mykytam replied:
Yes, it is, thanks

kh0ma commented on home-data/src/main/java/com/softserveinc/ita/homeproject/homedata/entity/User.java:
```diff
@@ -15,6 +15,7 @@
 import javax.persistence.Table;
 import javax.persistence.JoinTable;
 import javax.persistence.JoinColumn;
+import javax.persistence.FetchType;
```
Isn't this one redundant?

mykytam replied:
Yes, it is, thanks

kh0ma commented on home-service/src/main/java/com/softserveinc/ita/homeproject/service/constants/Permissions.java:
```diff
@@ -0,0 +1,9 @@
+package com.softserveinc.ita.homeproject.service.constants;
+
+public interface Permissions {
+    String createUserPermission = "hasAuthority('CREATE_USER_PERMISSION')";
```
Please follow java constant naming convention https://www.oracle.com/java/technologies/javase/codeconventions-namingconventions.html

mykytam replied:
Ok, I will do, thanks

kh0ma commented on home-service/src/main/java/com/softserveinc/ita/homeproject/service/constants/Roles.java:
```diff
@@ -0,0 +1,8 @@
+package com.softserveinc.ita.homeproject.service.constants;
+
+public interface Roles {
+    String userRole = "USER";
```
Please follow java constant naming convention https://www.oracle.com/java/technologies/javase/codeconventions-namingconventions.html

mykytam replied:
Ok, I will do, thanks

kh0ma commented on home-service/src/main/java/com/softserveinc/ita/homeproject/service/dto/CreateUserDto.java:
```diff
@@ -0,0 +1,16 @@
+@Getter
+@Setter
+@SuperBuilder
+public class CreateUserDto extends BaseUserServiceDto{
```
Does it really worth creating a base dto for users? If it is please rename `BaseUserDto`

mykytam replied:
We can make it without base dto. I just wanted to make more abstraction. If it's redundant, I can get rid of it

kh0ma commented on home-service/src/main/java/com/softserveinc/ita/homeproject/service/exceptions/ApiException.java:
```diff
@@ -0,0 +1,13 @@
+
+import org.springframework.http.HttpStatus;
+
+public class ApiException extends Exception{
```
Why do we have `Api...` in service? Why is it checked, do we really need to make it checked?

mykytam replied:
Ok, understood the issue, thanks

kh0ma commented on home-service/src/main/java/com/softserveinc/ita/homeproject/service/exceptions/NotFoundException.java:
```diff
@@ -0,0 +1,10 @@
+package com.softserveinc.ita.homeproject.service.exceptions;
```
I don't see any purpose in providing plural package naming. Can we make it singular?

mykytam replied:
Sure we can, thanks

kh0ma commented on home-service/src/main/java/com/softserveinc/ita/homeproject/service/exceptions/ApiException.java:
```diff
@@ -0,0 +1,13 @@
+
+    private HttpStatus status;
+
+    public ApiException (HttpStatus status, String msg) {
```
As it is a service exception no information about the presentation layer should appear here.

mykytam replied:
Ok, noted, thanks

kh0ma commented on home-service/src/main/java/com/softserveinc/ita/homeproject/service/impl/UserServiceImpl.java:
```diff
@@ -0,0 +1,85 @@
+            userRepository.save(toUpdate);
+            return conversionService.convert(toUpdate, ReadUserDto.class);
+        } else {
+            throw new NotFoundException(HttpStatus.NOT_FOUND, "User with this Id is not found");
```
Make error message more informative. Include given Id into it.

mykytam replied:
Ok, I will

mykytam committed:
Issue#35: Users CRUD
Delete redundant imports
Fix naming according to conventions
Delete redundant BaseUserServiceDto class
Change exception package naming
Make exception's messages more understandable

mykytam committed:
Issue#35: Users CRUD
Changes in open api spec - paging
Add pagination to users query
Add fields to User entity
Add crete and update dates to service

mykytam committed:
Issue#35: Users CRUD
Change contacts variable type

mykytam committed:
Issue#35: Users CRUD
Add pagination for get all operation

mykytam committed:
Issue#35: Users CRUD
Change contacts type in DTOs

mykytam committed:
Issue#35: Users CRUD
Change type of contacts in OpenApi spec

mykytam committed:
Issue#35: Users CRUD
Change exception naming
Delete http status from errors

mykytam committed:
Merge remote-tracking branch 'origin/dev' into Issue#35-Users-CRUD

mykytam committed:
Issue#35: Users CRUD
Change code due to new OpenApi config

mykytam committed:
Issue#35: Users CRUD
Service DTOs merged into one general UserDto

mykytam committed:
Issue#35: Users CRUD
Rename DTO to more general

mykytam committed:
Issue#35: Users CRUD
Change package naming

mykytam committed:
Issue#35: Users CRUD
Change variable naming

mykytam committed:
Merge remote-tracking branch 'origin/dev' into Issue#35-Users-CRUD

# Conflicts:
#	home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/exception/HomeException.java
#	home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/exception/NotFoundException.java
